src/coreclr/vm/typekey.h

```diff
@@ -253,6 +277,19 @@ class TypeKey
                 if (pKey1->u.asFnPtr.m_pRetAndArgTypes[i] != pKey2->u.asFnPtr.m_pRetAndArgTypes[i])
                     return FALSE;
             }
+
+            // For unmanaged function pointers, the signature must match.
+            if (pKey1->u.asFnPtr.m_sigLen != pKey2->u.asFnPtr.m_sigLen)
+            {
+                return FALSE;
+            }
+
+            if (pKey1->u.asFnPtr.m_sigLen > 0)
+            {
+                if (memcmp(pKey1->u.asFnPtr.m_sig, pKey2->u.asFnPtr.m_sig, pKey1->u.asFnPtr.m_sigLen))
```

Comment:
This will make two identical function pointers used in two different models to be treated as not equal. That does not sound right.

Reply:
This is an open question I have on "ZapSig". Currently that code serializes\deserializes a function pointer but does not include the calling conventions today so it would already have this problem but is made worse here by also not saving the signature. Not sure how "ZapSig" is used today.

Also note that having the type desc hold a reference to the signature here helps with lazy-processing of the custom mods and having a signature is compatible with existing managed code in the internal "Signature" class which in this PR helps with that lazy processing. However, we could eagerly load all of the custom mods and avoid having have a reference to the signature, as well as the Module which in this PR was added to support the lazy-processing.

Reply:
ZapSig is signature used to encode references from AOT code in R2R images. I do not think that what you have here is a ZapSig.

Reply:
No, it's not -- just trying to point out other areas of code that lookup a type descriptor for a function pointer, and since a function pointer should include the calling conventions for identity that other code may or may not be broken with function pointers today.

Also It's actually the debugger, not ZapSig, that doesn't support calling conventions today:
https://github.com/dotnet/runtime/blob/03e0187a1a912f38e718c00614201dd20fa1bff1/src/coreclr/vm/eedbginterfaceimpl.cpp#L886-L887
https://github.com/dotnet/runtime/blob/03e0187a1a912f38e718c00614201dd20fa1bff1/src/coreclr/vm/eedbginterfaceimpl.cpp#L1011-L1012
https://github.com/dotnet/runtime/blob/c299002ce9b5dad911a665ffae246c80dccf29f7/src/coreclr/debug/daccess/dacdbiimpl.cpp#L2370-L2373

ZapSig doesn't support custom mods today:
https://github.com/dotnet/runtime/blob/89dc073529aadfa3f36bcf7ed029c0a3b59d357e/src/coreclr/vm/zapsig.cpp#L58-L63

Reply:
> This will make two identical function pointers used in two different models to be treated as not equal. That does not sound right.

The latest code now does a full comparison against all custom modifiers without doing a memcmp (unless in the same module).

Note that in MetadataLoadContext, only `.Equals()` works as expected; ReferenceEquals() has not been supported - even before function pointer support was added.